feat: add endpoint to create an offer

HijoDeHonor opened the pull request:
- Implemented POST endpoint '/offers)' to create a new publication
- Added corresponding route, controller, service, and repository
- Created unit tests for the new endpoint, service, and repository
- Added text constants for better maintainability
- Improved error handling and validation

HijoDeHonor committed:
feat: add endpoint to create an offer

- Implemented POST endpoint '/offers)' to create a new publication
- Added corresponding route, controller, service, and repository
- Created unit tests for the new endpoint, service, and repository
- Added text constants for better maintainability
- Improved error handling and validation

PiewDev commented on src/offer/offerRepository.js:
```diff
@@ -1,59 +1,100 @@
+  async create (userName, offer, request) {
+    try {
+      const newOffer = await this.mySQLConnection.executeTransaction(async (connection) => {
+        const offerId = uuidv4();
```
this id should come from the front

HijoDeHonor committed:
fix: add additional verifications and update tests

- Added verifications to ensure data integrity and correctness
- Updated unit tests to reflect the new verifications and ensure comprehensive coverage

HijoDeHonor committed:
fix: add additional verifications and update tests

- Added verifications to ensure data integrity and correctness
- Updated unit tests to reflect the new verifications and ensure comprehensive coverage

PiewDev commented on src/offer/offerService.js:
```diff
@@ -1,12 +1,38 @@
+    this.inventoryRepository = inventoryRepository;
   }
 
+  create = async (userName, offer, request, id) => {
```
The parameters are normally written in order of importance, in this case Id is the most important because it is a unique identifier, the ideal would be: id, userName, offer, request

PiewDev commented on src/offer/offerService.js:
```diff
@@ -1,12 +1,38 @@
+    }
+    for (const item of offer) {
+      const { name, Quantity } = item;
+      const actualQuantity = await this.inventoryRepository.getQuantity(userName, name);
```
Maybe at this point it is not the best idea to deconstruct item, it is more readable to do item.prop

PiewDev commented on src/offer/offerService.js:
```diff
@@ -1,12 +1,38 @@
+    }
+    for (const item of offer) {
+      const { name, Quantity } = item;
+      const isDelete = await this.inventoryRepository.removeItemToUser(userName, name, Quantity);
```
Maybe at this point it is not the best idea to deconstruct item, it is more readable to do item.prop

PiewDev commented on src/offer/offerService.js:
I would have the user's items removed in the create transaction in the repository

PiewDev commented on src/inventory/inventoryRepository.js:
```diff
@@ -64,4 +65,69 @@ export class InventoryRepository {
+        `, [userName, item]
+        );
+        return res.affectedRows > 0;
+      } else {
```
This else is unnecesary

PiewDev commented on src/inventory/inventoryRepository.js:
```diff
@@ -64,4 +65,69 @@ export class InventoryRepository {
+        WHERE user_userName = ? AND item_Name = ?
+        `, [userName, name]
+      );
+      if (!rows.length === 0) {
```
if (rows.length === 0)*

PiewDev commented on src/inventory/inventoryRepository.js:
```diff
@@ -64,4 +65,69 @@ export class InventoryRepository {
+        `, [userName, name]
+      );
+      if (!rows.length === 0) {
+        return false;
```
Here you are returning false when the function is expected to return a number, if that item is not found the quantity you have is 0, you should return 0

PiewDev commented on src/inventory/inventoryRepository.js:
```diff
@@ -64,4 +65,69 @@ export class InventoryRepository {
+      );
+      if (!rows.length === 0) {
+        return false;
+      } return rows[0].Quantity;
```
add a line break

PiewDev reviewed (changes requested):
There are several important changes to make

HijoDeHonor committed:
refactor to accomplish the pr requirements

HijoDeHonor committed:
Merge branch 'develop' into feature/36-create-offer-endpoint
